ibc/docs/ics23.md

```diff
@@ -0,0 +1,89 @@
+
+See: [`cosmos/ics23` #152](https://github.com/cosmos/ics23/issues/152) and [`cosmos/ics23` #153](https://github.com/cosmos/ics23/pull/153ß) for the details of the changes made to allow for `ExclusionProof` verification.
+
+### Custom SMT `ProofSpec`
+
+The `ProofSpec` type in `cosmos/ics23` is used to define:
+
+1. The steps needed to verify a proof
+2. The hash functions used
+3. Node prefixes
+4. Etc...
+
+The `ProofSpec` is then passed into the verification functions in order to verify a proof instead of having to interact with the tree itself. This is useful as proofs must be verified via an (IBC) light client, and as such being able to verify a proof without reconstructing a tree is much more memory efficient.
+
+As the SMT used by Pocket Network only stores hashed values by default, the IBC store uses the `WithValueHasher(nil)` option which stores the source value (as raw bytes) in the tree. The following `ProofSpec` was created to support this:
+
+```go
+smtSpec *ics23.ProofSpec = &ics23.ProofSpec{
+	LeafSpec: &ics23.LeafOp{
+		Hash:         ics23.HashOp_SHA256,
+		PrehashKey:   ics23.HashOp_SHA256,
+		PrehashValue: ics23.HashOp_NO_HASH,
+		Length:       ics23.LengthOp_NO_PREFIX,
+		Prefix:       []byte{0},
+	},
+	InnerSpec: &ics23.InnerSpec{
+		ChildOrder:      []int32{0, 1},
+		ChildSize:       32,
+		MinPrefixLength: 1,
+		MaxPrefixLength: 1,
+		EmptyChild:      make([]byte, 32),
+		Hash:            ics23.HashOp_SHA256,
+	},
+	MaxDepth:                   256,
+	PrehashKeyBeforeComparison: true,
+}
+```
+
+The main difference from the `cosmos/ics23` `SmtSpec` object is that the `PrehashValue` field is set to not hash values before hashing the key-value pair.
+
+### Converting `SparseMerkleProof` to `CommitmentProof`
+
+In order to convert the proofs generated by the SMT into a serialisable proof used by `cosmos/ics23`, the `SideNodes` field of the `SparseMerkleProof` must be converted into a list of `InnerOp` types which define the order of the hashes. The order of the hashes is important as depending on whether the next hash is the left or right neighbour of the current hash, they will be hashed in a different order, ultimately creating a different root hash. This conversion allows the verification to produce the same root hash as the SMT would have produced when verifying the proof.
+
+As `SparseMerkleProof` objects represent both inclusion and exclusion proofs as defined in the [JMT whitepaper][jmt]. The conversion step will convert the SMT proof into either an `ExistenceProof` or `ExclusionProof` as defined in `cosmos/ics23`.
+
+### Proof Verification
+
+Membership proofs are verified as follows:
+
+1. Use the key-value pair to generate a leaf hash
+2. Hash the leaf with the `SideNodes` found in the `path` field of the `ExistenceProof` to generate the root hash
+3. Compare the root hash with the one provided and expect them to be identical
+
+Non-membership proofs are verified as follows:
```

Comment:
@h5law I don't know why I'm struggling to understand this documentation, but I put together this mermaid diagram. Can you edit it based on your understanding?

I still feel like I have a gap in my knowledge somewhere.

```mermaid

---
title: Exclusion Proof 
---
flowchart TD
    I["Proof,Key"]
    NMD{"proof.NonMembershipLeafData == nil ?"}
    KP1["actualPath = sha256(key) \n actualValue = placeholderNode"]
    KP2["actualPath = ProvidedKeyHash \n actualValue = ProvidedValueHash"]
    Compare{"ComputedRootHash == ProvidedRootHash ?"}
    R1{"actualValue == placeholderNode ?\n\n Tree matches - exclusion \n requires empty node for path(key)"}
    R2{"actualValue == providedValueHash ?\n\n Tree does not match - exclusion \n requires provided value at the path given"}    
    EV["Exclusion Prove VALID"]
    EI["Exclusion Prove INVALID"]

    I --> NMD
    NMD -- Yes --> KP1
    NMD -- No --> KP2

    KP1 -- Key,Value --> Compare
    KP2 -- Key,Value --> Compare
    
    Compare -- Yes --> R1
    Compare -- No --> R2

    R1 -- Yes --> EV
    R1 -- No --> EI
    R2 -- Yes --> EV
    R2 -- No --> EI    
```

Reply:
Have updated the docs with an accurate diagram